subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -529,242 +600,231 @@ func (h *HivedAlgorithm) findVirtualCell(
-			klog.Infof("Cells bound: %v and %v", vc.GetName(), pc.GetName())
-			if vc.GetParent() == nil {
-				break
-			}
-			vc = vc.GetParent().(*VirtualCell)
-			pc = pc.GetParent().(*PhysicalCell)
-		}
+// generatePodScheduleResult writes the scheduling result into a PodScheduleResult.
+func generatePodScheduleResult(
+	groupPhysicalPlacement map[int32][]CellList,
+	groupVirtualPlacement map[int32][]CellList,
+	currentGpuNum int32,
+	currentPodIndex int32,
+	suggestedNodes []string,
+	pod *core.Pod) internal.PodScheduleResult {
 
-		// set priority
-		setPriority(virtualCell, priority, unlimitedLevel)
-		// first hold the physical cell of the preassigned (set its priority to regularPriority)
-		if !preassignedBound {
-			setPriority(pac.GetPhysicalCell(), regularPriority, unlimitedLevel)
-		}
-		// set the priority of the physical cell to regularNonPreassignedPriority
-		// so that other cells inside the preassigned one can still be used by opportunistic pods
-		if virtualCell != pac {
-			setSelfAndParentPriority(physicalCell, regularNonPreassignedPriority, unlimitedLevel)
+	if groupPhysicalPlacement == nil {
+		return internal.PodScheduleResult{
+			PodWaitInfo: &internal.PodWaitInfo{
+				FailedNodeReasons: map[string]string{},
+			},
 		}
 	} else {
-		setPriority(physicalCell, opportunisticPriority, unlimitedLevel)
-	}
-}
-
-// confirmReleasedCell releases an allocated cell by destroying cell bindings and reset priorities.
-func confirmReleasedCell(c *PhysicalCell) {
-	var preassignedPhysical *PhysicalCell
-	if vc := c.GetVirtualCell(); vc != nil {
-		// destroy bindings
-		boundVirtual := vc
-		for !boundVirtual.GetPhysicalCell().IsReserved() {
-			boundPhysical := boundVirtual.GetPhysicalCell()
-			klog.Infof("Cells unbound: %v and %v", boundVirtual.GetName(), boundPhysical.GetName())
-			boundPhysical.SetVirtualCell(nil)
-			boundVirtual.SetPhysicalCell(nil)
-			if boundVirtual.GetParent() == nil {
-				break
-			} else {
-				unbindParent := true
-				for _, cc := range boundVirtual.GetParent().GetChildren() {
-					if child := cc.(*VirtualCell); child.GetPhysicalCell() != nil {
-						unbindParent = false
-						break
-					}
-				}
-				if !unbindParent {
-					break
-				}
-				boundVirtual = boundVirtual.GetParent().(*VirtualCell)
+		affinityGroupBindInfo, selectedNode, selectedGpuIndices := generateAffinityGroupBindInfo(
+			groupPhysicalPlacement, groupVirtualPlacement, currentGpuNum, currentPodIndex)
+		chain := string(groupPhysicalPlacement[currentGpuNum][currentPodIndex][0].GetChain())
+		if !common.StringsContains(suggestedNodes, selectedNode) {
+			panic(fmt.Sprintf("[%v]: node %v picked by algorithm but not in K8S candidates",
+				internal.Key(pod), selectedNode))
+		}
+		// collect preemption victims
+		preemptionVictims := common.NewSet()
+		for _, gpu := range groupPhysicalPlacement[currentGpuNum][currentPodIndex] {
+			pGpu := gpu.(*PhysicalCell)
+			if pGpu.HasPod() {
+				preemptionVictims.Add(pGpu.GetPods()[0])
 			}
```

Comment:
Return all victim Pods inside the victim affinity group to do gang preemption. So that different allocated affinity groups have not placement overlap/conflict.